src/disk_io_thread.cpp

```diff
@@ -3108,33 +3076,24 @@ namespace libtorrent
 
 		TORRENT_ASSERT((j->flags & disk_io_job::in_progress) || !j->storage);
 
-		// if there are at least 3 threads, there's a hasher thread
-		// and the hash jobs go into a separate queue
-		// see set_num_threads()
-		if (m_num_threads > 3 && j->action == disk_io_job::hash)
+		job_queue& q = queue_for_job(j);
+		q.m_queued_jobs.push_back(j);
+		// if we literally have 0 disk threads, we have to execute the jobs
+		// immediately. If add job is called internally by the disk_io_thread,
+		// we need to defer executing it. We only want the top level to loop
+		// over the job queue (as is done below)
+		if (pool_for_job(j).max_threads() == 0 && user_add)
 		{
-			m_queued_hash_jobs.push_back(j);
-		}
-		else
-		{
-			m_queued_jobs.push_back(j);
-			// if we literally have 0 disk threads, we have to execute the jobs
-			// immediately. If add job is called internally by the disk_io_thread,
-			// we need to defer executing it. We only want the top level to loop
-			// over the job queue (as is done below)
-			if (m_num_threads == 0 && user_add)
-			{
-				l.unlock();
-				immediate_execute();
-			}
+			l.unlock();
+			immediate_execute();
```

Comment:
Now that I think about this, it's really subtle. If `pool_for_job(j).max_threads() == 0` then it implies that there are no worker threads and we are running on the network thread. Since only the network thread changes the max threads there's no opportunity for a race.